project_task_personal_stage_auto_fold/models/project_task.py

```diff
@@ -0,0 +1,42 @@
+from odoo import models
+
+
+class Task(models.Model):
+    _inherit = "project.task"
+
+    def write(self, vals):
+        result = super().write(vals)
+        if "stage_id" in vals and vals.get("stage_id"):
+            if self.env["project.task.type"].browse(vals.get("stage_id")).is_closed:
+                self._fold_personal_stage_task()
+        return result
+
+    def _fold_personal_stage_task(self):
+        personal_tasks_not_folded = (
+            self.env["project.task.stage.personal"]
+            .sudo()
+            .search(
+                [
+                    ("task_id", "in", self.ids),
+                    ("user_id", "=", self.env.user.id),
+                    ("stage_id.fold", "=", False),
+                ]
+            )
+        )
+        fold_stage = (
+            self.env["project.task.type"]
+            .sudo()
+            .search(
+                [
+                    ("user_id", "=", self.env.user.id),
+                    "|",
+                    ("fold", "=", True),
+                    ("is_closed", "=", True),
+                ],
+            )
+            .sorted(lambda ptt: ptt.is_closed, reverse=True)
+        )
+
+        for personal_not_folded_task in personal_tasks_not_folded:
+            if fold_stage:
+                personal_not_folded_task.stage_id = fold_stage[0]
```

Comment:
This code contains a couple of `sudo()` calls that I think we could avoid. Also it uses the `is_closed` field in the wrong model, as explained above.

Also it doesn't take into account that you can close a task by moving it to "done" or to "cancelled". It's a bit tricky to get that 100% right, but there's some work that can be done to at least try it.

Besides, `project.task` already includes some helpers to easily move the personal stage, so we can reuse them.

I think this code will do. IMHO it's simpler and more accurate:

```suggestion
        # Find all folded personal stages
        folded_stages = (
            self.env["project.task.type"]
            .search(
                [
                    ("user_id", "=", self.env.user.id),
                    ("fold", "=", True),
                ],
            )
        )
        if not folded_stages:
            return
        # Allow to find personal stage with same name as shared stage
        stage_by_name = {stage.name: stage for stage in folded_stages}
        # Apply best matching personal stage
        for task in self:
            task.personal_stage_type_id = stage_by_name.get(
                task.stage_id.name,
                folded_stages[0],
            )
```

Reply:
Your suggestion works fine and it's better. 
I'm just going to add the sorted for `is_closed` field first on the `project.task.type` search.